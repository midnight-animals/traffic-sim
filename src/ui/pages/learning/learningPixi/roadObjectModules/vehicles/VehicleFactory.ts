// import { VehicleBuilder } from "./VehicleBuilder";
// import { VehicleDirector } from "./VehicleDirector";
// import { carConstants } from "./cars/carConstants";

// export class VehicleFactory {
//   public makeCar() {
//     const vehicleDirector = new VehicleDirector();
//     const vehicleBuilder = new VehicleBuilder();
//     const car = vehicleDirector.constructCar(vehicleBuilder);
//     car.position;
//   }
// }
