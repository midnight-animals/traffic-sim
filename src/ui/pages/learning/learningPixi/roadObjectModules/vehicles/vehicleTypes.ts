export type VehicleType = "truck" | "car" | "motorcycle";
