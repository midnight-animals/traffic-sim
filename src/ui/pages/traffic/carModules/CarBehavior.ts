import { ICar } from "./carTypes";

/**
 * TODO
 * - Add more behaviors
 *   - change lanes
 *     - if other lane (seems to be) moving faster
 *     - if car in front is moving too slowly
 *     - (if needs to turn)
 */
export class CarBehavior {
  public isAccelerating: boolean;
  public isChangingLanes: boolean;
  public isParking: boolean;

  constructor(private car: ICar) {}

  public setIsAccelerating(is: boolean): void {
    this.isAccelerating = is;
  }
  public setIsChangingLanes(is: boolean): void {
    this.isChangingLanes = is;
  }
  public setIsParking(is: boolean): void {
    this.isParking = is;
  }

  public wantsToAccelerate(): boolean {
    const wants = Math.random() >= 0.95;
    if (wants) {
      this.setIsParking(true);
    }
    return wants;
  }

  public wantsToChangeLane(): boolean {
    const wants = Math.random() >= 0.95;
    if (wants) {
      this.setIsChangingLanes(true);
    }
    return wants;
  }

  public wantsToPark(): boolean {
    const wants = Math.random() >= 0.95;
    if (wants) {
      this.setIsParking(true);
    }
    return wants;
  }
}
