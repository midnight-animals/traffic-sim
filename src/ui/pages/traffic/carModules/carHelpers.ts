export const CAR_HEIGHT = 20;
export const CAR_HEIGHT_FACTOR = 0.3;
export const CAR_GAP = 30;
export const CAR_WIDTH = 50;
export const CAR_ACCELERATION = 3;
export const CHANGE_LANE_SPEED = 3;

/**
 * Cars within range in the front and in the rear = "nearby"
 */
export const NEARBY_VALUE = CAR_WIDTH / 2;

export function getCarHeight(laneRect: DOMRect) {
  return CAR_HEIGHT || laneRect.height * CAR_HEIGHT_FACTOR;
}
