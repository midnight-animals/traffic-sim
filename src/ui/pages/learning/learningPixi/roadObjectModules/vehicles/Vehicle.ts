import { Rectangle } from "pixi.js";
import { IRoadObject } from "../IRoadObject";
import { VehicleType } from "./vehicleTypes";
import { VehicleBehavior } from "../../../../traffic/carModules/VehicleBehavior";

export class Vehicle extends IRoadObject {
  public behavior: VehicleBehavior;

  constructor(
    public type: VehicleType,
    public rect: Rectangle,
    public speed: number
  ) {
    super();
  }

  // Add methods for Vehicle class here
  setBehavior(behavior: VehicleBehavior) {
    this.behavior = behavior;
  }
}
