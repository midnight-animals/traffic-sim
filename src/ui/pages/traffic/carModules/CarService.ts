import {
  getCenterOfCarInLane,
  getLaneRect
} from "../../learning/learningPixi/roadModules/roadhelpers";
import { RoadConfig } from "../trafficTypes";
import { CHANGE_LANE_SPEED, NEARBY_VALUE } from "./carHelpers";
import { ICar } from "./carTypes";

export class CarService {
  private context: CanvasRenderingContext2D;

  constructor(public canvas: HTMLCanvasElement) {
    this.context = canvas.getContext("2d");
  }

  // Function to update the position of a car
  accelerate(car: ICar): void {
    car.x += car.speed;
    if (car.x > this.canvas.width) {
      car.x = -20; // Reset the car's position to the start when it reaches the end of the canvas
    }
  }

  slowDown(car: ICar): void {
    throw new Error("Method not implemented.");
  }

  move(car: ICar, roadConfig: RoadConfig): void {
    this.accelerate(car);
    // car.behavior.wantsToChangeLane() && this.changeLane(car, roadConfig);
    car.behavior.wantsToChangeLane();
    const toLane = this.getLaneToChangeTo(car, roadConfig);
    const canChange = this.canChangeLanes(car, toLane, roadConfig.cars);
    if (canChange) {
      if (car.behavior.isChangingLanes) {
        this.changeLanes(car, toLane);
      }
    }
  }

  getNearbyCars(targetCar: ICar, cars: ICar[]): ICar[] {
    const nearbyCars = cars.filter((car) => {
      if (car.id === targetCar.id) return false;

      const nearbyRear = Math.abs(targetCar.x - car.x) <= NEARBY_VALUE;
      const frontCar = car.x + car.width;
      const frontTargetCar = targetCar.x + targetCar.width;
      const nearbyFront = Math.abs(frontTargetCar - frontCar) <= NEARBY_VALUE;
      const nearby = nearbyRear || nearbyFront;
      return nearby;
    });
    return nearbyCars;
  }

  /**
   * TODO:
   * - check if `to` lane exists
   */
  canChangeLanes(car: ICar, toLane: number, cars: ICar[]): boolean {
    if (car.lane === toLane) return false;
    if (car.x < 0) return false;
    car.lane; /*?*/
    toLane; /*?*/
    const nearbyCars = this.getNearbyCars(car, cars);
    if (nearbyCars.length === 0) {
      return true;
    }

    const nearbyCarsOnToLane = nearbyCars.filter((car) => car.lane === toLane);
    if (nearbyCarsOnToLane.length === 0) {
      return true;
    }
    // const mapped = cars.map((car) => ({ ...car, behavior: null }));
    // /* prettier-ignore */ console.log('>>>> _ >>>> ~ mapped:', mapped);

    return false;
  }

  changeLanes(car: ICar, to: number): void {
    const laneRect = getLaneRect(this.canvas, to);
    const center = getCenterOfCarInLane(to, laneRect);
    const sign = car.lane < to ? 1 : -1;

    if (sign > 0) {
      if (car.y >= center) {
        car.lane = to;
        car.behavior.setIsChangingLanes(false);
        return; // Already in the correct lane
      }
    } else {
      if (car.y <= center) {
        car.lane = to;
        car.behavior.setIsChangingLanes(false);
        return; // Already in the correct lane
      }
    }

    car.y += CHANGE_LANE_SPEED * sign;

    if (car.y > this.canvas.height) {
      car.y = -20; // Reset the car's position to the start when it reaches the end of the canvas
    }
  }

  getLaneToChangeTo(car: ICar, roadConfig: RoadConfig): number {
    const currentLane = car.lane;
    const numOfLanes = roadConfig.lanes.length;
    // Cases at borders
    if (currentLane === numOfLanes) {
      return currentLane - 1;
    } else if (currentLane === 0) {
      return 1;
    }

    return 0;
  }
}
