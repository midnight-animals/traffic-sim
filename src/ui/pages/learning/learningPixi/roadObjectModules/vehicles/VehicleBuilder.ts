import { Rectangle } from "pixi.js";
import { Vehicle } from "./Vehicle";
import { VehicleType } from "./vehicleTypes";

export class VehicleBuilder {
  type: VehicleType;
  rect = new Rectangle(0, 0, 100, 100);
  speed = 0;

  setType(type: VehicleType): VehicleBuilder {
    this.type = type;
    return this;
  }

  setSpeed(speed: number): VehicleBuilder {
    this.speed = speed;
    return this;
  }

  setRect(rect: Rectangle): VehicleBuilder {
    this.rect = rect;
    return this;
  }

  build(): Vehicle {
    this.checkField("type");
    this.checkField("rect");
    this.checkField("speed");

    return new Vehicle(this.type, this.rect, this.speed);
  }

  public checkField(fieldName: keyof typeof VehicleBuilder.prototype) {
    if (this[fieldName] == null) {
      throw new Error(`Missing ${fieldName} for Vehicle`);
    }
  }
}
