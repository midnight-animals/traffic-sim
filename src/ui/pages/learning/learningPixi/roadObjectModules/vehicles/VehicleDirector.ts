import { VehicleBuilder } from "./VehicleBuilder";
import { carConstants } from "./cars/carConstants";

export class VehicleDirector {
  public constructCar(builder: VehicleBuilder) {
    return builder.setType("car").setRect(carConstants.rect).build();
  }
}
