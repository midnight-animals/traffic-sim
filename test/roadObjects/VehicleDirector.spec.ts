import { VehicleDirector } from "../../src/ui/pages/learning/learningPixi/roadObjectModules/vehicles/VehicleDirector";
import { VehicleBuilder } from "../../src/ui/pages/learning/learningPixi/roadObjectModules/vehicles/VehicleBuilder";
import { VehicleBehavior } from "../../src/ui/pages/traffic/carModules/VehicleBehavior";

describe("VehicleDirector", () => {
  const vehicleDirector = new VehicleDirector();
  const vehicleBuilder = new VehicleBuilder();
  describe("VehicleDirector", () => {
    it("Car", () => {
      const car = vehicleDirector.constructCar(vehicleBuilder);
      expect(car).toMatchSnapshot();
    });
    it.only("Car with behavior ISSUE-xBqg8BnF", () => {
      const car = vehicleDirector.constructCar(vehicleBuilder);
      const behavior = new VehicleBehavior();
      car.setBehavior(behavior);
      expect(car).toMatchSnapshot();
    });
  });
});
