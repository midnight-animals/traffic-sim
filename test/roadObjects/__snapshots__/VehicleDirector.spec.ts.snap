// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VehicleDirector VehicleDirector Car 1`] = `
Vehicle {
  "rect": Rectangle {
    "height": 100,
    "type": 1,
    "width": 200,
    "x": 0,
    "y": 0,
  },
  "speed": 0,
  "type": "car",
}
`;

exports[`VehicleDirector VehicleDirector Car 2`] = `
Vehicle {
  "rect": Rectangle {
    "height": 100,
    "type": 1,
    "width": 200,
    "x": 0,
    "y": 0,
  },
  "speed": 0,
  "type": "car",
}
`;

exports[`VehicleDirector VehicleDirector Car with behavior ISSUE-xBqg8BnF 1`] = `
Vehicle {
  "behavior": CarBehavior {},
  "rect": Rectangle {
    "height": 100,
    "type": 1,
    "width": 200,
    "x": 0,
    "y": 0,
  },
  "speed": 0,
  "type": "car",
}
`;
